Batteries/Data/OrderedAssocList.lean

```diff
@@ -0,0 +1,1018 @@
+/-
+Copyright (c) 2023 Lean FRO, LLC. All rights reserved.
+Released under Apache 2.0 license as described in the file LICENSE.
+Authors: Kim Morrison
+-/
+import Batteries.Data.AssocList
+import Batteries.Classes.Order
+
+/-!
+# Ordered association lists
+
+`OrderedAssocList` is a wrapper around `AssocList`,
+with the additional invariant that the keys are in strictly increasing order.
+
+As a consequence, an `OrderedAssocList` is determined by the lookup function `l[x]?`, that is
+`(∀ a, l₁[a]? = l₂[a]?) → l₁ = l₂`
+and this makes proving identities between operations much easier than with `AssocList`.
+
+As the operations on `OrderedAssocList` are defined in terms of `AssocList`,
+they evaluate in the kernel, but inherit the `O(n)` time complexity
+(i.e. are not suitable for large data). Their advantage relative to an ordered map is the
+simple extensionality property above.
+
+We will later add another wrapper requiring that a "default" value does not appear,
+so e.g. finitely supported functions can be uniquely represented.
+
+The main operations defined are:
+* `l[x]?`, which linearly searches the list, stopping if the keys get too large.
+* `insert`, which inserts a new key-value pair, maintaining the order invariant.
+* `filterMapVal f`, for `f : α → β → Option γ`, which applies a function to values,
+  dropping some values.
+* `merge f` for `f : α → Option β → Option γ → Option δ` which merges two lists,
+  dropping some values. It runs in time `O(l₁.length + l₂.length)`.
+-/
+
+namespace Batteries
+
+/-!
+We first define some predicates and operations in the `AssocList` namespace.
+
+* `keysOrdered cmp l` asserts that the keys of an `l : AssocList` are strictly increasing
+  with respect to a comparator `cmp`.
+* `ltHeadKey? cmp a l` asserts that `a` is less than (according to `cmp`) the first key of `l`,
+  or that `l` is empty.
+-/
+
+namespace AssocList
+
+/-- The head key of an `AssocList`, or `none` if the list is empty. -/
+def headKey? (l : AssocList α β) : Option α :=
+  match l with
+  | .nil => none
+  | .cons a _ _ => some a
+
+@[simp] theorem headKey?_nil : headKey? (.nil : AssocList α β) = none := rfl
+@[simp] theorem headKey?_cons : headKey? (.cons a b t) = some a := rfl
+
+/--
+The condition that an element is less than the first key of an `AssocList`, or that list is empty.
+-/
+abbrev ltHeadKey? (cmp : α → α → Ordering) (a : α) (t : AssocList α β) : Prop :=
```

Comment:
Makes more sense as `Bool` valued. Something about the `?` makes me unhappy here but it's hard to think of a more accurate alternative.